### ISSUE TITLE
Workaround for JSON type not being supported by DDL simulation logic

### ISSUE THREAD

**Issue by vojtapol** [jOOQ/jOOQ#10978]
### Your question:
We would like to have jOOQ generator run from our Flyway migrations. One of our migrations has this statement in it:

```sql
CREATE TABLE `page` (
  `id`     CHAR(22) COLLATE `ascii_bin` NOT NULL PRIMARY KEY,
  `name`   VARCHAR(190)                 NOT NULL,
  `schema` JSON                         NOT NULL
);
```

This throws an exception because `JSON` is not supported (by the parser? or the H2 database?) for now. We don't really need jOOQ to know that this column is a `JSON`. It would be okay if it thought that it was `TEXT` for example. 

Are there any special constructs to use to have the real database execute different lines than the jOOQ parser?

This is what I have in mind:
```sql
CREATE TABLE `page` (
  `id`     CHAR(22) COLLATE `ascii_bin` NOT NULL PRIMARY KEY,
  `name`   VARCHAR(190)                 NOT NULL,
  /*!    `schema` JSON                  NOT NULL */
  /*jooq `schema` TEXT                  NOT NULL */
);
```

### Versions:

- jOOQ: 3.13.2
- Java: 11
- Database (include vendor): MySQL

**Comment by lukaseder** [comment]
We have a syntax that is ignored by jOOQ's parser, but none that is parsed by jOOQ but ignored by the databases. Here's are some examples of how to have the jOOQ parser ignore some syntax:
https://www.jooq.org/doc/latest/manual/sql-building/dsl-context/custom-settings/settings-parser/

**Comment by lukaseder** [comment]
Of course, in your specific case, you could just search-replace the SQL input prior to passing it to the jOOQ parser. Searching `\bJSON\b` replacing by `TEXT` should work quite well

**Comment by lukaseder** [comment]
This too hacky?

```sql
CREATE TABLE `page` (
  `id` CHAR(22) COLLATE `ascii_bin` NOT NULL PRIMARY KEY
  ,`name` VARCHAR(190) NOT NULL
  /* [jooq ignore start] */
  ,`schema` JSON NOT NULL
  /* [jooq ignore stop] */
  /* [jooq ignore start] */ -- /* [jooq ignore stop] */ ,`schema` TEXT NOT NULL
);
```

**Comment by vojtapol** [comment]
@lukaseder I think I can work with that. Thank you!

In addition to `[jooq ignore start]` it would be nice to add support for:

```
/* [jooq]
ONLY EXECUTED BY JOOQ;
*/
```

Those two in combination should allow for working around almost any parser issues.

**Comment by vojtapol** [comment]
> Of course, in your specific case, you could just search-replace the SQL input prior to passing it to the jOOQ parser. Searching `\bJSON\b` replacing by `TEXT` should work quite well

Is there a way to have the parser execute some pre-processing code before it starts parsing (maybe defined in `pom.xml`), or would we have to take care of that ourselves?

**Comment by lukaseder** [comment]
> In addition to `[jooq ignore start]` it would be nice to add support for:

Yes, that's why I renamed this issue here. An explicit syntax is a bit nicer and more manageable than my suggested hack where the `--` token is ignored only by jOOQ. Until then, that hack will work

> Is there a way to have the parser execute some pre-processing code before it starts parsing (maybe defined in `pom.xml`), or would we have to take care of that ourselves?

We could implement that for some trivial cases, but people will soon ask for more complex search/replace features, when there are Maven plugins just for that, e.g. the `com.google.code.maven-replacer-plugin:maven-replacer-plugin`

**Comment by lukaseder** [comment]
The syntax will be, for consistency reasons:

```sql
/* [jooq only start]
XYZ
   [jooq only stop] */
```

This allows for adding the tokens also in the presence of other comments, e.g.:

```sql
/* [jooq only start]
XYZ /* still a comment */
   [jooq only stop] also still a comment */
```

**Comment by lukaseder** [comment]
Could make sense on single line comments as well:

```sql
-- [jooq only start] XYZ [jooq only end] still a comment
```

This would translate to (in jOOQ's parser):

```sql
/* */ XYZ /* still a comment */
```

Would be easier to implement if we already had a tokenizer: https://github.com/jOOQ/jOOQ/issues/10671

**Comment by lukaseder** [comment]
Indeed, I think this depends on implementing the tokenizer first. It's harder to implement than the `[jooq ignore start]` tokens

**Comment by lukaseder** [comment]
Meanwhile, I'll document the "hack" in the manual. It's still useful...

**Comment by vojtapol** [comment]
@lukaseder Thank you Lukas. I think those solutions sound fantastic!
